### ISSUE TITLE
Where is Macro Value Stored?

### ISSUE THREAD

**Issue by wonderb** [chris-kent/Live-Input-Macros#11]
I am trying to get the value selected in a dropdown list via the REST API, but I cannot tell where the value is being stored.  
I tried in the properties, but it's not there.  

I noticed in the HTML that is output from a page properties report macro that the macro's ID is LIM-DropdownList-<contentid>-<macroid>, but I was not able to access it using that as the ID either.

I'm not sure how to get the macro hash due to the unfortunate lack of good documentation of the REST API.

I'm running Confluence version 5.7.5.

Help would be appreciated since we really need to use the dropdown and get values via the API.

**Comment by wonderb** [comment]
Of course as soon as I post the question, I figure out the answer.  :)

If anyone else is looking, here's how to get the value.

Instead of using /rest/api/content/<contentID>?expand=body._storage_
Use /rest/api/content/<contentID>?expand=body.**_view_**

From there you can get the stored value by parsing the value found between these div tags:
`<div style="display:none" id="LIM-DropdownList-<contentID>-<macroID>\"></div>`

**Comment by wonderb** [comment]
So I still need your help if possible.  I also need to be able to _write_ a value using the REST API.  I can't use body.view to write. :(

Where is the selected value of the dropdown stored so that I can try to figure out how to write to wherever it is?
Thank you in advance!

**Comment by chris-kent** [comment]
Hi there,

I don't have a complete REST API.  All I do is post new selected values to the servlet.
The get is not needed, as the page refresh does this via the macro.

The actual data is stored in the confluence database, so if you have access to SQL commands and the db, then that would be your cleanest solution.  If not, then I can give you REST commands to post data.

Tell me which you prefer and also which macros you need to update, ie is it only the dropdown?

Cheers

**Comment by wonderb** [comment]
Hi Chris,

Thank you for the quick response!  Unfortunately, I do not have access to the Confluence db.  I know the REST API commands for creating and updating a page (found [here](https://docs.atlassian.com/atlassian-confluence/REST/6.6.0/)).  Do you know of a different way to post specifically for the macro selected values?

This would be for dropdowns, checkboxes, radio buttons, and text boxes.

Thank you!

**Comment by chris-kent** [comment]
Hi,
I have started to document this, please see this article: https://kcintegrations.atlassian.net/wiki/spaces/PLUG/pages/149553155/REST+API+Version+1.0 

HTH
Chris

**Comment by chris-kent** [comment]
Looks like the url is wrong, it's either my documentation or your code.
Please send me the url you are using and I'll confirm
Chris

**Comment by wonderb** [comment]
For security purposes, I can't put the real URL, but here's kind of what it looks like:
https://my-wiki.mydomain.com/plugins/servlet/liveinput/dropdownListServlet

I used what was in the documentation link you posted.  The only difference is I'm using https instead of http.  I don't know if that makes a difference or not.

**Comment by chris-kent** [comment]
I've got an idea. Use the macros as designed on a Confluence page. Turn on browser debugging, and check the Netwotk traffic and see what the url is when you set a dropdown using the Confluence page.

This will also give you the parameters you need too.

**Comment by wonderb** [comment]
I don't know if this is what you're looking for, but when I change the value of the drop down, I see this (edited for security):
`dropdown changed: context:LIM-DropdownList-MyPageID-MyMacroID selected:My Option 3 batch.js:1234:5678`
and
`checkbox changed: context:LIM-CheckBox-MyPageID-MyMacroID item:My Option 2 val:true 1 batch.js:1234:567`

I also see this, which I don't think is related but may give more info:
`DEPRECATED - Inline dialog constructor has been deprecated and will be removed in a future release. blah blah blah...
https://my-wiki.mydomain.com/s/numbersandletters-CDN/en_GB/1234/8numbersandletters.8/numbersandletters/_/download/contextbatch/js/atl.confluence.plugins.pagetree-desktop,main,viewcontent,atl.general,page,plugin.quick.comment.pre/batch.js`

numbersandletters being a bunch of numbers and letters that I don't recognize

**Comment by wonderb** [comment]
Ah, I see this under Network in debugging (which, now I see that's what you said to do :P )
https://my-wiki.mydomain.com/plugins/servlet/liveinput/dropdownListServlet

which is exactly what I used, so that's not the issue, I don't think.

**Comment by chris-kent** [comment]
So, what exactly is the command you are using?
Remember this is a POST command, not the default GET.

**Comment by wonderb** [comment]
I figured it out! Woo hoo! :)

From looking at the debugger, how the syntax is and parameters are passed, I saw how I needed to pass them.
I'm used to passing Json via the REST API.  Doing it the new (for me) way does not use the REST API, it uses the plugin's servlet.

For example: 
REST API would pass json something like this: 
`{"option":"Option1", "pageId":"MyPageID", "macroId":"MyMacroID"}`

Using the servlet, I pass this:
`option=Option1&pageId=MyPageID&macroId=MyMacroID`

I don't know if it makes a difference or not, but I copied most of the request headers from the debugger and changed them from how I normally use the REST API as well.
All I know is that it worked. :)
Thank you so much for your help!
